src/VisualStudio/Core/Def/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -111,79 +138,176 @@ private async ValueTask ProcessWorkspaceChangeAsync(CancellationToken cancellati
             }
 
             // Wait until all project updates have happened before processing the next batch.
             await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            // now that we've reported this data, record that we've done so so that we don't report the same data again in the future.
+            foreach (var data in changedData)
+                _documentToLastReportedData[data.DocumentId] = data;
+
+            // Now, check the documents we've stored against the changed projects to see if they're no longer around.
+            // If so, dump what we have.
+            if (latestSolution != null)
+            {
+                foreach (var (documentId, _) in _documentToLastReportedData)
+                {
+                    if (!latestSolution.ContainsDocument(documentId))
+                        _documentToLastReportedData.TryRemove(documentId, out _);
+                }
+            }
         }
 
-        private static void AddFilteredInfos(ImmutableSegmentedList<DesignerAttributeData> data, ArrayBuilder<DesignerAttributeData> filteredData)
+        private CodeAnalysis.Solution? AddChangedData(
+            ImmutableSegmentedList<(CodeAnalysis.Solution? solution, DesignerAttributeData? data)> dataList,
+            ArrayBuilder<DesignerAttributeData> changedData)
         {
-            using var _ = PooledHashSet<DocumentId>.GetInstance(out var seenDocumentIds);
+            using var _1 = PooledHashSet<DocumentId>.GetInstance(out var seenDocumentIds);
+            using var _2 = ArrayBuilder<DesignerAttributeData>.GetInstance(out var latestData);
+
+            CodeAnalysis.Solution? lastSolution = null;
+
+            for (var i = dataList.Count - 1; i >= 0; i--)
+            {
+                // go in reverse order so that results about the same document only take the later value.
+                var (solution, data) = dataList[i];
+
+                if (data != null)
+                {
+                    if (seenDocumentIds.Add(data.Value.DocumentId))
+                        latestData.Add(data.Value);
+                }
+
+                lastSolution ??= solution;
+            }
 
-            // Walk the list of designer items in reverse, and skip any items for a project once
-            // we've already seen it once.  That way, we're only reporting the most up to date
-            // information for a project, and we're skipping the stale information.
-            for (var i = data.Count - 1; i >= 0; i--)
+            foreach (var data in latestData)
             {
-                var info = data[i];
-                if (seenDocumentIds.Add(info.DocumentId))
-                    filteredData.Add(info);
+                if (!_documentToLastReportedData.TryGetValue(data.DocumentId, out var existingData) ||
+                    existingData.Category != data.Category)
+                {
+                    changedData.Add(data);
```

Comment:
so it looks like we compare the changed docIds (aka everything that came from the dataList that wasn't null) to the last reported docIds and only trigger notifications for ones that are different which seems fine.

But where do we compare the last reported docs that are no longer present in the new data (aka the designer attribute was removed)?